### PR TITLE
Gaussian elimination final

### PR DESCRIPTION
**Pull request by andriybriliant**
Line 460 to line 505 disabled flags because we didn't really know how to split function gaussian_diagonal into helper functions so that it complies with clang-tidy requirements and make it look not ugly at the same time. Also removed enum class axis because it's not neccessary in the next project.

### CODE REVIEW

File: algebra2/gaussian_elimination/gaussian_elimination.hpp [hunk 1]
```diff
@@ -20,267 +22,322 @@ namespace algorithms::gaussian_elimination {
...
+            std::vector<I> copy(m.begin(), m.end());
+            ::ranges::matrix_view<I, LP> mat{
+                copy, m.extent(0), m.extent(1), LP{}};
+            return std::pair{std::move(copy), mat};
+        } else {
+            auto fracs{convert_to_vector_of_fractions(m)};
+            ranges::matrix_view<fraction<I>, LP> matrix_of_fracs{
+                fracs, m.extent(0), m.extent(1), LP{}};
+            return std::pair{std::move(fracs), matrix_of_fracs};
+        }
```
Comment (lemkis) [inline algebra2/gaussian_elimination/gaussian_elimination.hpp:270]: move to matrix.cxx and rename it (this function can be easily generalized - you just change type of matrix so it is matrix util sth like `auto to_type()`

File: algebra2/gaussian_elimination/gaussian_elimination.hpp [hunk 2]
```diff
@@ -20,267 +22,322 @@ namespace algorithms::gaussian_elimination {
...
         identity_vector.reserve(dimensions);
-        
-        for(int i = 0; i < dimensions; i++){
+
+        for (std::size_t i = 0; i < dimensions; i++) {
             identity_vector[i] = T{0};
         }
-        for(int i = 0; i < dimensions; i += rows+1){
+        for (std::size_t i = 0; i < dimensions; i += rows + 1) {
             identity_vector[i] = T{1};
```
Comment (lemkis) [inline algebra2/gaussian_elimination/gaussian_elimination.hpp:288]: already implemented -check if works as you require

File: algebra2/gaussian_elimination/gaussian_elimination.hpp [hunk 3]
```diff
@@ -20,267 +22,322 @@ namespace algorithms::gaussian_elimination {
...
+    /*
+       description:
+           combines two matrices into one augmented matrix 
+           where matrix m1 is on the right and matrix m2 is on the left of the combined matrix
+   */
+    template <typename T, typename LP>
+    auto combine_matrices(ranges::matrix_view<T, LP> m1,
+                          ranges::matrix_view<T, LP> m2) {
+        const std::size_t combined_dims = m1.number_of_columns() + m2.number_of_columns();
+        const std::size_t vector_size = combined_dims * m1.number_of_rows();
```
Comment (lemkis) [inline algebra2/gaussian_elimination/gaussian_elimination.hpp:306]: move and rename to `utils::matrix::concat`

File: algebra2/gaussian_elimination/gaussian_elimination.hpp [hunk 4]
```diff
@@ -20,267 +22,322 @@ namespace algorithms::gaussian_elimination {
...
         description:
             splits one matrix in half and returns the rightmost half
     */
     template <typename T, typename LP>
-    auto split_matrix(ranges::matrix_view<T, LP> m){
-        int rows = m.extent(0),
-            cols = m.extent(1),
-            dimensions = rows * cols;
+    auto split_matrix(ranges::matrix_view<T, LP> m) {
+        const std::size_t rows = m.number_of_rows();
```
Comment (lemkis) [inline algebra2/gaussian_elimination/gaussian_elimination.hpp:328]: move to matrix.cxx

File: algebra2/gaussian_elimination/gaussian_elimination.hpp [hunk 5]
```diff
@@ -290,273 +347,281 @@ namespace algorithms::gaussian_elimination {
...
-    enum class axis {
-        rows,
-        columns,
-    };
 
-    enum class reducted_form {
-        echelon,
-        diagonal
-    };
+    enum class reducted_form : std::uint8_t { echelon, diagonal };
```
Comment (lemkis) [inline algebra2/gaussian_elimination/gaussian_elimination.hpp:357]: change diagonal to echelon_reducted

File: algebra2/gaussian_elimination/gaussian_elimination.hpp [hunk 6]
```diff
@@ -290,273 +347,281 @@ namespace algorithms::gaussian_elimination {
...
-                            }
+        for (std::size_t i = 0; i < rows; i++) {
+            if (m[i, i].numerator != 0) {
+                for (std::size_t k = i + 1; k < rows; k++) {
+                    if (m[k, k].numerator != 0) {
+                        factor = m[i, k] / m[k, k];
+                        if (factor.numerator != 0) {
+                            subtract(m, i, k, factor);
+                            result.reduction_steps.push_back(std::format(
+                                "R{} - {} * R{}", k + 1, factor, i + 1));
```
Comment (lemkis) [inline algebra2/gaussian_elimination/gaussian_elimination.hpp:472]: simplify logic - hard to read

File: algebra2/gaussian_elimination/gaussian_elimination.hpp [hunk 7]
```diff
@@ -290,273 +347,281 @@ namespace algorithms::gaussian_elimination {
...
-            return gaussian_elimiantion_alg(m, axis::rows, reducted_form::diagonal).determinant_m;
-        }
+    auto determinant(ranges::matrix_view<T, LP> m)
+        -> std::expected<fraction<int>, error> {
+        const std::size_t rows = m.number_of_rows();
+        const std::size_t cols = m.number_of_columns();
+
+        if (rows != cols) { return std::unexpected(error::not_square); }
+        return gaussian_elimiantion_alg(m, reducted_form::diagonal)
+            .determinant_m;
```
Comment (lemkis) [inline algebra2/gaussian_elimination/gaussian_elimination.hpp:598]: this should be in namespace utils::matrix  
similarly with other matrix methods

File: algebra2/gaussian_elimination/tests.hpp
Comment (lemkis) [inline algebra2/gaussian_elimination/tests.hpp:1]: move to the tests directory. Make examples tests  or  move examples to main.cxx file

**Comment by andriybriliant** [comment]
When using `find ./algebra2 ./discrete_math/ -iname '*.cxx' -o -iname '*.ixx' | xargs clang-tidy-18  -p ./build/debug/ --extra-arg=-std=c++26` not every file is compiled for some reason, but when using gnu++26 instead of c++26 everything works.

**Pull request closed without merging**